jbpm-flow/src/main/java/org/jbpm/process/core/impl/ObjectCloner.java

```diff
@@ -0,0 +1,479 @@
+    }
+
+    @SuppressWarnings({"rawtypes", "unchecked"})
+    private static Map deepCloneMap(Map<?, ?> object) {
+        Map result = null;
+        Optional<Constructor<?>> constructor = getCollectionConstructor(object.getClass());
+
+        if (constructor.isPresent()) {
+            try {
+
+                result = (Map) constructor.get().newInstance();
+            } catch (ReflectiveOperationException e) {
+                logger.warn("Unexpected exception invoking default constructor of type {}", object.getClass());
+            }
+        }
+
+        if (result == null) {
+            result = new HashMap<>();
+        }
+        for (Entry item : object.entrySet()) {
+            result.put(item.getKey(), clone(item.getValue()));
+        }
+        return result;
+    }
+
+    private static Object lightCloneCollectionMap(Object object, Class<?> paramType) {
+        return getCollectionConstructor(object.getClass(), paramType).map(c -> {
+            try {
+                return c.newInstance(object);
+            } catch (ReflectiveOperationException e) {
+                logger.warn("Unexpected exception invoking constructor {} with object {}", c, object);
+                return object;
+            }
+        }).orElse(object);
+
+    }
+    
+    private static <T> Optional<Constructor<?>> getCollectionConstructor (Class<?> collectionClass, Class<?>... paramTypes) {
+        if (unmutableClasses.contains(collectionClass)) {
+            return Optional.empty();
+        }
+        try {
+            return Optional.of(collectionClass.getConstructor(paramTypes));
+        } catch (NoSuchMethodException ex) {
+            logger.debug("Copy constructor not found for type {}", collectionClass, ex);
+            unmutableClasses.add(collectionClass);
+            return Optional.empty();
+        }
+    }
+
+    private static Object cloneObject(Object object) {
+        Class<?> objectClass = object.getClass();
+
+        if (object instanceof Cloneable) {
+            try {
+                return objectClass.getMethod("clone").invoke(object);
+            } catch (ReflectiveOperationException ex) {
+                logger.warn("Failing to call clone in a cloneable object!!", ex);
+                return object;
+            }
+        } else {
+            return cloneWithConstructor(object, objectClass);
+        }
+    }
+
+    private static Object cloneWithConstructor(Object object, Class<?> objectClass) {
+        return constructorExecutorMap.computeIfAbsent(objectClass, ObjectCloner::getConstructorExecutor).apply(object);
+    }
+
+    private static UnaryOperator<Object> getConstructorExecutor(Class<?> objectClass) {
+        ConstructorInfo constructorInfo = new ConstructorInfo(objectClass);
+        if (constructorInfo.getCopyConstructor() != null) {
+            return new CopyConstructorExecutor(constructorInfo.getCopyConstructor());
+        } else {
```

Comment:
this "else" is not strictly necessary if returns in the "if" (one indentation level can be removed)

Reply:
ok, code rearranged to avoid too many returns in same method